### PR TITLE
upgrade to plugin-framework

### PR DESCRIPTION
**Pull request by megan07**
If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: mmv1/third_party/terraform/main.go.erb [hunk 1]
```diff
@@ -2,11 +2,71 @@
...
-		ProviderFunc: google.Provider})
+	var debug bool
+
+	flag.BoolVar(&debug, "debug", false, "set to true to run the provider with support for debuggers like delve")
+	flag.Parse()
+
+	// plugin framework provider
+	downgradedFrameworkProvider, err := tf6to5server.DowngradeServer(
+		context.Background(),
+		providerserver.NewProtocol6(google.New(version)),
```
Comment (bflad) [inline mmv1/third_party/terraform/main.go.erb:36]: Drive-by comment: Rather than serving the framework provider via protocol version 6 then downgrading to protocol version 5, the mux server can use the framework provider on protocol version 5 directly instead: https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/providerserver#NewProtocol5

```go
providers := []func() tfprotov5.ProviderServer{
	providerserver.NewProtocol5(google.New(version)),
	google.Provider().GRPCProvider,
}
```

File: mmv1/third_party/terraform/utils/framework_provider.go.erb [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+                },
+            },
+            "resource_manager_v3_custom_endpoint": &schema.StringAttribute{
+                Optional:     true,
+                Validators: []validator.String{
+                    CustomEndpointValidator(),
+                },
+            },
+            <% unless version == "ga" -%>
+            "runtimecconfig_custom_endpoint": &schema.StringAttribute{
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/utils/framework_provider.go.erb:182]: Typo here!

(copying review comments over from the stacked PR)

File: mmv1/third_party/terraform/data_sources/data_source_dns_record_set.go [hunk 3]
```diff
@@ -1,86 +1,145 @@
...
+	Type        types.String `tfsdk:"type"`
+	Project     types.String `tfsdk:"project"`
+}
 
-			"ttl": {
-				Type:     schema.TypeInt,
-				Computed: true,
-			},
+func (d *GoogleDnsRecordSetDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_dns_record_set"
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/data_sources/data_source_dns_record_set.go:39]: (Moving review comments across from stacked PR)

This is a source of an issue - TypeName needs to be "the full data source type, including the provider type prefix and an underscore. For example, examplecloud_thing." ([from docs](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v1.1.1/datasource#MetadataResponse)).

Even when using the [google-beta provider the types still only start with `google_`](https://github.com/hashicorp/terraform-provider-google-beta/blob/4a7212618248e0dcd8dcb94cf4a3d9bbd04757bf/google-beta/provider.go#L941) not `google-beta_`

However, the value of `req.ProviderTypeName` is "google-beta" so the provider is sending a `datasource.MetadataResponse` with `google-beta_dns_record_set` as the type name

Bet that's causing an issue somewhere, but I'm not sure where!

**Comment by SarahFrench** [comment]
The failure after merging in that PR is

```
Step #13 - "tfv-head": /go/pkg/mod/github.com/modular-magician/terraform-provider-google@v1.20.1-0.20230131143652-39c685bc42bb/google/test_utils.go:210:21: undefined: MuxedProviders
```
🤔

**Comment by megan07** [comment]
> The failure after merging in that PR is
> 
> ```
> Step #13 - "tfv-head": /go/pkg/mod/github.com/modular-magician/terraform-provider-google@v1.20.1-0.20230131143652-39c685bc42bb/google/test_utils.go:210:21: undefined: MuxedProviders
> ```
> 
> But it's defined in mmv1/third_party/terraform/framework_utils/framework_provider_test.go and both files are within package google 🤔

I think its because its defined in a `*_test.go` file it's not exported. I'll rename/rearrange some files to clean that up a bit.

File: mmv1/third_party/terraform/framework_utils/framework_config.go.erb [hunk 4]
```diff
@@ -0,0 +1,744 @@
...
+			return *creds
+		}
+
+		creds, err := googleoauth.CredentialsFromJSON(ctx, []byte(contents), clientScopes...)
+		if err != nil {
+			diags.AddError("unable to parse credentials", err.Error())
+			return googleoauth.Credentials{}
+		}
+
+		log.Printf("[INFO] Authenticating using configured Google JSON 'credentials'...")
```
Comment (bendbennett) [inline mmv1/third_party/terraform/framework_utils/framework_config.go.erb:682]: Believe you could use [tflog](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-log/tflog#Info) in place of `log.Printf`.

For instance:
```
tflog.Info(ctx, "Authenticating using configured Google JSON 'credentials'...")
```

File: mmv1/third_party/terraform/framework_utils/framework_transport.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+			reqHeaders.Set("X-Goog-User-Project", "")
+		} else {
+			// Pass the project into this fn instead of parsing it from the URL because
+			// both project names and URLs can have colons in them.
+			reqHeaders.Set("X-Goog-User-Project", project)
+		}
+	}
+
+	if timeout == 0 {
+		timeout = time.Duration(1) * time.Hour
```
Comment (bendbennett) [inline mmv1/third_party/terraform/framework_utils/framework_transport.go:38]:
```suggestion
		timeout = time.Hour
```

File: mmv1/third_party/terraform/tests/data_source_dns_record_set_test.go.erb [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+data "google_dns_record_set" "rs" {
+  managed_zone = google_dns_record_set.rs.managed_zone
+  name         = google_dns_record_set.rs.name
+  type         = google_dns_record_set.rs.type
+}
+`, managedZoneName, zoneName, recordSetName)
+}
+
+// Framework checkdestroy
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/tests/data_source_dns_record_set_test.go.erb:70]: It'd be useful to make this comment similar to the one above the other function for managed zones

```suggestion
// testAccCheckDnsRecordSetDestroyProducerFramework is the framework version of the generated testAccCheckDnsRecordSetDestroyProducer
// when we automate this, we'll use the automated version and can get rid of this
```

File: tpgtools/templates/provider_dcl_endpoints.go.tmpl [hunk 7]
```diff
@@ -63,6 +62,32 @@ func configureDCLProvider(provider *schema.Provider) {
 {{- end}}
 }
 
+func HandleDCLProviderDefaults(d *schema.ResourceData) {
+{{- range $index, $pkg := . }}
+{{- if $pkg.ShouldWriteProductBasePath }}
```
Comment (SarahFrench) [inline tpgtools/templates/provider_dcl_endpoints.go.tmpl:67]: Glad to see this could be made dynamic! 🙌

File: mmv1/third_party/terraform/main.go.erb [hunk 8]
```diff
@@ -2,11 +2,57 @@
...
 )
 
+var (
+	// these will be set by the goreleaser configuration
+	// to appropriate values for the compiled binary
+	version string = "dev"
+
+	// goreleaser can also pass the specific commit if you want
+	// commit  string = ""
+)
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/main.go.erb:23]: Does anything need to be changed with [version/version.go](https://github.com/hashicorp/terraform-provider-google/blob/main/version/version.go) and the release workflows to work with the muxed provider?

Comment (SarahFrench) [reply mmv1/third_party/terraform/main.go.erb:23]: Just surfacing info : Here's where `version.ProviderVersion` is used in the code:

https://github.com/GoogleCloudPlatform/magic-modules/blob/3456e2c8a2882be6ccb585401fba57a91ea2b3ac/mmv1/third_party/terraform/utils/provider.go.erb#L536-L540

And here's where the goreleaser process overrides the value of that variable in the release process:

https://github.com/hashicorp/terraform-provider-google/blob/5bc7515c1a5b8e2248222be77386a894414717a8/.goreleaser.yml#L31-L32

---

Could you use `version.ProviderVersion` instead of this new version string? Or is this the new way things are done for framework plugins?

Comment (SarahFrench) [reply mmv1/third_party/terraform/main.go.erb:23]: Copying over details from a Slack conversation:

The new variables `version` and `commit` need to be reflected in the goreleaser file to let them be over-written. [This is the line in the framework scaffold repo that allows their values to be overwritten vid ldsflags argument.](https://github.com/hashicorp/terraform-provider-scaffolding-framework/blob/b6f0e7fef1edffef4eb01a1b5e31122edf7d7bf1/.goreleaser.yml#L17)

In the old SDK code we get info about what version is being released via a version package and a var called `ProviderVersion`. [In goreleaser files for SDK providers the ldflags argument is different, and overwrites that var instead.](https://github.com/hashicorp/terraform-provider-google/blob/5bc7515c1a5b8e2248222be77386a894414717a8/.goreleaser.yml#L32)

If our muxed provider is going to have [this code unchanged](https://github.com/GoogleCloudPlatform/magic-modules/blob/3456e2c8a2882be6ccb585401fba57a91ea2b3ac/mmv1/third_party/terraform/utils/provider.go.erb#L536-L540) and still using `version.ProviderVersion` and also use the new `version` var in the google package, then we'd need to create a new combined version of the ldflags argument like this (**though I'd want to check it manually first**):

```
- '-s -w -X main.version={{.Version}} -X main.commit={{.Commit}} -X github.com/hashicorp/terraform-provider-google/version.ProviderVersion={{.Version}}'
```

File: mmv1/third_party/terraform/data_sources/data_source_dns_record_set.go (hunk 3 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/data_sources/data_source_dns_record_set.go:39]: Addressed in 267067ae70cdf59a0a93e3a2033b248ecad9aafa

**Review by bendbennett — APPROVED** [review]

File: mmv1/third_party/terraform/main.go.erb (hunk 8 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/main.go.erb]: Addressed in 1ce4eb6 - I've manually confirmed that it creates the framework user agent OK

**Review by SarahFrench — APPROVED** [review]
Approved - I can use the build and use the muxed provider locally and acceptance tests pass, so although I don't 100% understand the muxing I'm happy to approve.
